### PR TITLE
Allow absolute links for refs with an identifier

### LINKED ISSUE #5148: HAPI-1094: Resource DocumentReference/urn:oid:1.2.3not found, specified in path: Provenance.entity.what
**Describe the bug**
We have an implementation that publishes FHIR resources that have been extracted from a CDA document. The implementation is following the IHE Profiles MXDE and QEDm. Due to this, we create a Provenance resource that points back to the CDA document which resides in an XDS Registry/Repository using our FHIR IHE MHD implementation. 

We have a variety of external resources such as Patients which are stored in our EMPI which is accessed by the PIXm IHE Profile. We do not see any errors related to any references to these from resource types such as Observations, Medications, or AllergyIntollerance so I can assume the configuration "allow_external_references" is working as expected. So the error in question seems to be specific to the Provenance resource and/or the "entity.what.reference" handling.

This is how we reference the XDS Document,

`"entity": [
                    {
                        "role": "source",
                        "what": {
                            "reference": "https://somehost:8443/mhd/r4/responder/DocumentReference/urn:oid:1.2.40.0.13.1.1.584109161.20230630175432428.38899",
                            "type": "DocumentReference",
                            "identifier": {
                                "use": "official",
                                "system": "urn:oid:2.16.840.1.113883.3.704.2.103.2.1.1.1.1.8",
                                "value": "urn:oid:1.2.40.0.13.1.1.584109161.20230630175432428.38899"
                            },
                            "display": "CDA Document in XDS Repository"
                        }
                    }
                ]`

Since our customer updated their FHIR server to v6.6.0 we are now seeing this error:
`{
    "resourceType": "OperationOutcome",
    "issue": [
        {
            "severity": "error",
            "code": "processing",
            "diagnostics": "HAPI-1094: Resource DocumentReference/urn:oid:1.2.40.0.13.1.1.584109161.20230630175432428.38899 not found, specified in path: Provenance.entity.what"
        }
    ]
}`

Here is the stack trace from the server:
`2023-07-31 11:26:07.403 [qtp709671086-22] DEBUG c.u.f.j.d.tx.HapiTransactionService [HapiTransactionService.java:237] Unexpected transaction exception. Will not be retried.
ca.uhn.fhir.rest.server.exceptions.InvalidRequestException: HAPI-1094: Resource DocumentReference/urn:oid:1.2.40.0.13.1.1.584109161.20230630175432428.38899 not found, specified in path: Provenance.entity.what
        at ca.uhn.fhir.jpa.dao.index.DaoResourceLinkResolver.findTargetResource(DaoResourceLinkResolver.java:127)
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.resolveTargetAndCreateResourceLinkOrReturnNull(SearchParamExtractorService.java:755)
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.extractResourceLinks(SearchParamExtractorService.java:610)
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.extractResourceLinks(SearchParamExtractorService.java:482)
        at ca.uhn.fhir.jpa.searchparam.extractor.SearchParamExtractorService.extractFromResource(SearchParamExtractorService.java:144)
        at ca.uhn.fhir.jpa.dao.index.SearchParamWithInlineReferencesExtractor.populateFromResource(SearchParamWithInlineReferencesExtractor.java:94)
        at ca.uhn.fhir.jpa.dao.BaseHapiFhirDao.updateEntity(BaseHapiFhirDao.java:1044)
        at ca.uhn.fhir.jpa.dao.BaseHapiFhirDao.updateInternal(BaseHapiFhirDao.java:1382)
        at ca.uhn.fhir.jpa.dao.BaseHapiFhirDao$$FastClassBySpringCGLIB$$741dcd55.invoke(<generated>)`

Locally, I have been using a FHIr server v5.7.2, I have not had any of the above issues. Only when I update to a later version (v6.6.0 to match the customer) I see issues.

- Is there a configuration I am missing?
- I checked the Provenance Resource specification, but I see nothing new on how the "entity.what.reference" should be populated... :/

**To Reproduce in v6.6.0**
1. Download the https://github.com/hapifhir/hapi-fhir-jpaserver-starter/tree/v6.6.0 
1. I have attached the application.yaml (as a text file) configuration files - the only change I made was to allow_external_references
2. run using mvn jetty:run
3. I have attached an example payload (test_bundle.txt), I have removed most resources from it so that it's very simple

**Expected behavior**
I expect the FHIR bundle to be published to the FHIR server 

**Environment (please complete the following information):**
 - HAPI FHIR 6.6.0 REST Server (FHIR Server; FHIR 4.0.1/R4)
 - OS: Running on mac 13.4.1 , but the same issue on Linux

Appreciated any help!

Thanks,
Matthew

[application.txt](https://github.com/hapifhir/hapi-fhir/files/12214064/application.txt)
[test_bundle.txt](https://github.com/hapifhir/hapi-fhir/files/12214083/test_bundle.txt)

### PR DESCRIPTION
**Pull request by jamesagnew**
Fixes #5148 and addresses a regression caused by #4736

### CODE REVIEW

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5295-fix-absolute-refs-with-identifier.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+type: fix
+issue: 5295
+title: "A regression in the HAPI FHIR 6.6.0 JPA server meant that absolute resource
+    references which also contained an identifier were rejected even if the server
+    was configured to allow absolute references. This has been correected."
```
Comment (KevinDougan-SmileCDR) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5295-fix-absolute-refs-with-identifier.yaml:6]:
```suggestion
    was configured to allow absolute references. This has been corrected."
```

Comment (dmuylwyk) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/5295-fix-absolute-refs-with-identifier.yaml:6]: **praise:** Thank you, @KevinDougan-SmileCDR, for correecting the correection.

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirResourceDaoR5PatchTest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class FhirResourceDaoR5PatchTest extends BaseJpaR5Test {
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirResourceDaoR5PatchTest.java:35]: **suggestion:** Remove this class from this PR.

This class is empty, and seems unrelated to the other changes. Did this come over from another branch by accident?

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/SearchParamExtractorService.java [hunk 3]
```diff
@@ -692,8 +693,20 @@ private void extractResourceLinks(
...
+		// can include URLs (e.g. token system URLs) and these really confuse
+		// the IdType parser because a conditional URL isn't actually a valid
+		// FHIR ID. So in order to truly determine whether we're dealing with
+		// an absolute reference, we strip the query part and reparse
+		// the reference.
+		int questionMarkIndex = nextId.getValue().indexOf('?');
+		if (questionMarkIndex != -1) {
+			IdType preQueryId = new IdType(nextId.getValue().substring(0, questionMarkIndex - 1));
+			baseUrl = preQueryId.getBaseUrl();
+		}
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/SearchParamExtractorService.java:708]: **question:** Is there a related test for this change?

The new `docref-test-bundle.json` in this PR doesn't include any conditional URLs.

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirSystemDaoTransactionR5Test.java [hunk 4]
```diff
@@ -503,6 +506,17 @@ public void testConditionalDeleteAndConditionalUpdateOnSameResource() {
 	}
 
 
+	@Test
+	public void testExternalReference() throws IOException {
+		myStorageSettings.setAllowExternalReferences(true);
+
+		Bundle input = loadResourceFromClasspath(Bundle.class, "docref-test-bundle.json");
+		Bundle output = mySystemDao.transaction(mySrd, input);
+		assertEquals(5, output.getEntry().size());
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirSystemDaoTransactionR5Test.java:515]: **question:** How does this test work?

The submitted Bundle has a single entry but the output from the transaction is expected to include 5 entries.

**Review by dmuylwyk — APPROVED** [review]
Approved with a few minor comments.

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirResourceDaoR5PatchTest.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirResourceDaoR5PatchTest.java:35]: Whoops.. so that's where that went. It was intended for another branch I was working on and I lost it! deleted here.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/SearchParamExtractorService.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/SearchParamExtractorService.java:708]: The issue is actually that line 695-696 in the (red) replaced code above was added a few months ago to deal with references that have identifiers, but it broke support for absolute references. Fixing it caused another test to fail which used conditionals, so this is addressing that other test.

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirSystemDaoTransactionR5Test.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirSystemDaoTransactionR5Test.java:515]: Whoops this test was actually not passing. Originally there were 5 resources in that bundle, but I removed the rest once I figured out which one was the MRP

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/SearchParamExtractorService.java (hunk 3 above)
Comment (dmuylwyk) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/SearchParamExtractorService.java:708]: Okay so long as there are existing tests. Thanks!

File: hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirSystemDaoTransactionR5Test.java (hunk 4 above)
Comment (dmuylwyk) [reply hapi-fhir-jpaserver-test-r5/src/test/java/ca/uhn/fhir/jpa/dao/r5/FhirSystemDaoTransactionR5Test.java]: Well, that's a relief. If I can't count to 5, I'm in trouble. : )
